File: typescript/src/aabb.ts
import { Hittable } from './model'
import { Ray } from './ray'

export class AABB {
  min: [number, number, number]
  max: [number, number, number]

  constructor(hittables: Hittable[]) {
    const mins = hittables.map((h) => h.getMinPoint())
    const maxs = hittables.map((h) => h.getMaxPoint())
    this.min = [
      Math.min(...mins.map((v) => v.x)),
      Math.min(...mins.map((v) => v.y)),
      Math.min(...mins.map((v) => v.z)),
    ]
    this.max = [
      Math.max(...maxs.map((v) => v.x)),
      Math.max(...maxs.map((v) => v.y)),
      Math.max(...maxs.map((v) => v.z)),
    ]
  }

  // https://tavianator.com/2015/ray_box_nan.html
  intersects(ray: Ray) {
    const origin = [ray.start.x, ray.start.y, ray.start.z]
    const invDir = [ray.inverseDirection.x, ray.inverseDirection.y, ray.inverseDirection.z]

    let t1 = (this.min[0] - origin[0]) * invDir[0];
    let t2 = (this.max[0] - origin[0]) * invDir[0];

    let tmin = Math.min(t1, t2);
    let tmax = Math.max(t1, t2);

    for (let i = 1; i < 3; i++) {
      t1 = (this.min[i] - origin[i]) * invDir[i]
      t2 = (this.max[i] - origin[i]) * invDir[i]
      tmin = Math.max(tmin, Math.min(t1, t2));
      tmax = Math.min(tmax, Math.max(t1, t2));
    }

    return tmax >= Math.max(tmin, 0);
  }
}

File: rust/src/aabb.rs
use crate::ray::Ray;
use nalgebra::Point3;

pub struct AABB {
    pub min: Point3<f64>,
    pub max: Point3<f64>,
}

impl AABB {
    pub fn new(min: Point3<f64>, max: Point3<f64>) -> AABB {
        AABB { min, max }
    }

    // https://tavianator.com/2015/ray_box_nan.html
    pub fn intersects(&self, r: &Ray, t_min: f64, t_max: f64) -> bool {
        let mut tmin = t_min;
        let mut tmax = t_max;
        for i in 0..3 {
            let inv_d = r.dir[i].recip();
            let t1 = (self.min[i] - r.origin[i]) * inv_d;
            let t2 = (self.max[i] - r.origin[i]) * inv_d;
            tmin = tmin.max(t1.min(t2));
            tmax = tmax.min(t1.max(t2));
            if tmax < tmin.max(0.0) {
                return false;
            }
        }
        true
    }

    pub fn get_surrounding_aabb(box0: &AABB, box1: &AABB) -> AABB {
        let small = Point3::new(
            f64::min(box0.min.x, box1.min.x),
            f64::min(box0.min.y, box1.min.y),
            f64::min(box0.min.z, box1.min.z),
        );
        let big = Point3::new(
            f64::max(box0.max.x, box1.max.x),
            f64::max(box0.max.y, box1.max.y),
            f64::max(box0.max.z, box1.max.z),
        );
        AABB::new(small, big)
    }
}
